### PR TITLE
Card Management System v1.4

### LINKED ISSUE #118: As a creator I can set a lesson's question and answer index.
So I can choose which core fields I want to use as the question and answer when I am attempting a lesson.

### LINKED ISSUE #151: UG command example error
**Describe the bug**
there's this example for `addCard` command that you gave for the `Country of Capitals` lesson (at least I hope its for that lesson I hope). 
image

**To Reproduce**
image
Type this command. Then open this lesson using `openLesson 1` command. then type this -
image

**Expected behavior**
Use appropriate and consistent example throughout UG for a particular lesson.

**Screenshots**
Given above

<hr>

**Reported by:** @rahulb99
**Severity:** Low

<sub>[original: nus-cs2103-AY1819S2/pe-dry-run#342]</sub>

### LINKED ISSUE #181: Label for lesson in lesson view does not display correctly when it is too long
**Describe the bug**
When a super long label for lesson card is added, it is not truncated or wrapped properly.

**To Reproduce**
1. Enter `addLesson n/Test Lesson q/Super duper long name which is never ever ending a/What's up`
2. Notice that the label exceeds the screen

**Expected behavior**
The label to be truncated.

**Screenshot**
exceeds

### LINKED ISSUE #182: Label for lesson in card view does not display correctly when it is long
**Describe the bug**
When a super long label for lesson card is added, it is not truncated or wrapped properly.

**To Reproduce**
1. Enter `addLesson n/Test Lesson q/Super duper long name which is never ever ending a/What's up`
2. Enter `openLesson x` where x is the index of the `Test Lesson`
2. Notice that the label exceeds the screen

**Expected behavior**
The label to be truncated.

**Screenshot**
exceeds4

### LINKED ISSUE #184: Unable to delete sample lesson
**Describe the bug**
When attempting to delete sample lesson an error message is thrown.

**To Reproduce**
1. Launch application for the first time with 0 lessons in '/data' folder.
2. Attempt to delete generated sample lesson with `deleteLesson 1` command.
3. Error thrown:
image

**Expected behavior**
Successful delete.

### LINKED ISSUE #187: Test value is duplicated as hint value upon `reload`
**Describe the bug**
After reloading the lessons, either by the `reload` command or upon exiting and then relaunching the app, a lesson's last test value is duplicated as a hint value.

**To Reproduce**
1. Launch application for the first time with 0 lessons in '/data' folder.
2. Enter the `reload` command or simply restart the app.
3. See lesson `Sample-Muscle-Anatomy`'s test value `Innervation` duplicated as hint value:
lesson_view

**Expected behavior**
Reloading does not cause duplication of test/hint values.

### PR DESCRIPTION
**Pull request by DigiPie**
### Enhance Card Management System (CMS)
This is a follow up to the PR: **Card Management System v1.4 (#173)** which was merged in milestone [**mid-v1.4**](https://github.com/CS2103-AY1819S2-W14-1/main/pull/177).

#### CMS enhancements 👍 
- Implements `setTest` command - use it to mark the 2 test values to be tested for a lesson (closes #118)
- Refactors `openLesson` command words to `edit` - for more intuitive usage
- Refactors `closeLesson` command word to `quit` - for more intuitive usage
- Refactors `reloadLessons` command word to `reload` - for easier usage
- Enhances `SampleDataUtil` - adds `SampleLessons` and `SampleCards` to replace hard-coded sample lesson generation

#### CMS GUI enhancements 🔢 
- Implements result box text colour change to red to indicate failed command
- Update `DarkTheme.css` and `LightTheme.css` files for better GUI
- Update `.fxml` files for better GUI

Lesson View |  Card View
:-------------------------:|:-------------------------:
lesson_view | card_view

Thanks @halcyoneee for helping with the legend footer!

#### Bug fixes 🐞 
- Fixes bug where a lesson's test value is duplicated as hint value (fixes #187)
- Fixes UI bug for Lesson and Flashcard card UI components (fixes #181, fixes #182)
- Fixes UI bug where command box text colour does not change to red to indicate failed command
- Fixes sample lesson delete bug - previously was unable to delete sample lesson, this has been rectified (fixes #184)

#### Refactoring and other improvements 🥇 
- Refactors `seedu.address.commons.util` classes with private constructors (therefore preventing instantiation):
- Writes JUnit tests for `seedu.address.commons.core` package to increase coverage to **99%*

#### Documentation ✍️ 
- Updates User Guide, Developer Guide and Personal Project Portfolio
- Rectifies inconsistent examples in User Guide (resolves #151)

### CODE REVIEW

File: src/main/java/seedu/address/commons/util/AppUtil.java [hunk 1]
```diff
@@ -10,6 +10,7 @@
  * A container for App specific utility functions
  */
 public class AppUtil {
+    private AppUtil() { }
```
Comment (eugenefdw) [inline src/main/java/seedu/address/commons/util/AppUtil.java]: These aren't really implementing singleton, a "proper" implementation would be "un-static-ing" all functions and only exposing a single static getInstance() for each class, then calling all functions/values through that instance. Not wrong as it helps with the coverage, but not accurate to call it singleton.

**Comment by eugenefdw** [comment]
The alternating colour in the list of cards/lessons being green might cause confusion with the green "Not tested" legend in the footer. A different colour scheme could be a brighter blue for tested, darker blue for untested, and a grey or a more dull colour for hints. Could be another colour, but given the way tested/untested are related to each other, related colours should be used.

File: src/main/java/seedu/address/commons/util/AppUtil.java (hunk 1 above)
Comment (DigiPie) [reply src/main/java/seedu/address/commons/util/AppUtil.java]: Hi @eugenefdw thanks for spotting this. I have read up on singleton-design and realised that what I am attempting to achieve is not singleton-design. I have updated the pull-request's description, Javadoc comments and code accordingly.

For more info see:
https://stackoverflow.com/a/2816181

**Comment by DigiPie** [comment]
> 
> 
> The alternating colour in the list of cards/lessons being green might cause confusion with the green "Not tested" legend in the footer. A different colour scheme could be a brighter blue for tested, darker blue for untested, and a grey or a more dull colour for hints. Could be another colour, but given the way tested/untested are related to each other, related colours should be used.

Thanks for your feedback @eugenefdw ! I have improved the colour scheme. Will upload screenshots soon.

**Pull request merged**